// kumva-sms-nodes/nodes/decision-maker.js
/**
 * DecisionMaker Node for Node-RED
 * Evaluates conditions and branches flow based on results
 * Supports multiple condition types: expression, global variable, contact existence, etc.
 */

module.exports = function(RED) {
    "use strict";

    function DecisionMakerNode(config) {
        RED.nodes.createNode(this, config);
        
        const node = this;
        this.conditionType = config.conditionType || "expression"; // expression, global, contact, input
        this.conditionMode = config.conditionMode || "binary"; // binary (yes/no) or multi
        this.conditions = config.conditions || [];
        this.expression = config.expression || "";
        this.globalVar = config.globalVar || "";
        this.contactId = config.contactId || "";

        node.on('input', function(msg) {
            try {
                let result = false;
                let evaluatedCondition = null;

                switch(node.conditionType) {
                    case 'expression':
                        evaluatedCondition = evaluateExpression(node.expression, msg);
                        result = Boolean(evaluatedCondition);
                        node.log(`Expression "${node.expression}" evaluated to: ${result}`);
                        break;

                    case 'global':
                        const globalValue = RED.nodes.getContext().global.get(node.globalVar);
                        evaluatedCondition = globalValue;
                        result = Boolean(globalValue);
                        node.log(`Global variable "${node.globalVar}" = ${globalValue}, result: ${result}`);
                        break;

                    case 'contact':
                        // Check if contact exists (simulated)
                        const contactId = node.contactId || msg.payload?.contactId || msg.contactId;
                        evaluatedCondition = contactId ? "exists" : "missing";
                        result = Boolean(contactId);
                        node.log(`Contact check for ID "${contactId}": ${result}`);
                        break;

                    case 'input':
                        // Check if required input exists
                        const inputValue = msg.payload || msg.input;
                        evaluatedCondition = inputValue;
                        result = Boolean(inputValue);
                        node.log(`Input check: ${result}`);
                        break;

                    case 'multi':
                        // Evaluate multiple conditions
                        const multiResults = evaluateMultipleConditions(node.conditions, msg, RED);
                        result = multiResults.result;
                        evaluatedCondition = multiResults.details;
                        node.log(`Multi-condition evaluation result: ${result}`);
                        break;

                    default:
                        node.error("Unknown condition type: " + node.conditionType);
                        return;
                }

                // Route based on result
                msg.decisionResult = result;
                msg.evaluatedCondition = evaluatedCondition;

                if (node.conditionMode === 'binary') {
                    // Binary mode: yes/no outputs
                    const outputs = [null, null];
                    outputs[result ? 0 : 1] = msg;
                    node.status({
                        fill: result ? "green" : "red",
                        shape: "dot",
                        text: result ? "TRUE" : "FALSE"
                    });
                    node.send(outputs);
                } else if (node.conditionMode === 'multi') {
                    // Multi-condition mode: send to matching condition output
                    const outputs = new Array(node.conditions.length + 1).fill(null);
                    
                    let matchedIndex = node.conditions.length; // default to "else" output
                    if (Array.isArray(evaluatedCondition)) {
                        for (let i = 0; i < evaluatedCondition.length; i++) {
                            if (evaluatedCondition[i]) {
                                matchedIndex = i;
                                break;
                            }
                        }
                    }
                    
                    outputs[matchedIndex] = msg;
                    node.status({
                        fill: "blue",
                        shape: "dot",
                        text: `condition ${matchedIndex}`
                    });
                    node.send(outputs);
                }

            } catch (err) {
                node.error("Error evaluating condition: " + err.message, msg);
                node.status({fill:"red", shape:"ring", text:"error"});
            }
        });

        function evaluateExpression(expr, msg) {
            try {
                // Create a safe evaluation context
                const context = {
                    msg: msg,
                    payload: msg.payload,
                    global: RED.nodes.getContext().global.get(),
                    flow: RED.nodes.getContext().flow.get(),
                    Math: Math,
                    Date: Date,
                    String: String,
                    Number: Number,
                    Boolean: Boolean
                };

                // Use Function constructor for safe evaluation
                const fn = new Function(...Object.keys(context), 'return ' + expr);
                return fn(...Object.values(context));
            } catch (err) {
                node.error("Expression evaluation error: " + err.message);
                return false;
            }
        }

        function evaluateMultipleConditions(conditions, msg, RED) {
            const results = [];
            
            conditions.forEach((condition, index) => {
                let condResult = false;

                if (condition.type === 'expression') {
                    condResult = Boolean(evaluateExpression(condition.expression, msg));
                } else if (condition.type === 'global') {
                    const value = RED.nodes.getContext().global.get(condition.variable);
                    condResult = Boolean(value);
                } else if (condition.type === 'comparison') {
                    // Simple comparison: field operator value
                    const field = getNestedValue(msg, condition.field);
                    condResult = performComparison(field, condition.operator, condition.value);
                }

                results.push(condResult);
            });

            // Return true if any condition matches
            return {
                result: results.some(r => r),
                details: results
            };
        }

        function getNestedValue(obj, path) {
            return path.split('.').reduce((current, prop) => current?.[prop], obj);
        }

        function performComparison(fieldValue, operator, compareValue) {
            switch(operator) {
                case 'eq': return fieldValue == compareValue;
                case 'ne': return fieldValue != compareValue;
                case 'gt': return fieldValue > compareValue;
                case 'lt': return fieldValue < compareValue;
                case 'gte': return fieldValue >= compareValue;
                case 'lte': return fieldValue <= compareValue;
                case 'contains': return String(fieldValue).includes(compareValue);
                case 'regex': return new RegExp(compareValue).test(fieldValue);
                default: return false;
            }
        }

        node.on('close', function() {
            node.status({});
        });
    }

    RED.nodes.registerType("decision-maker", DecisionMakerNode);
};
