// kumva-sms-nodes/nodes/send-sms.js
/**
 * SendSMS Node for Node-RED
 * Sends SMS messages via Africa's Talking
 */

module.exports = function(RED) {
    "use strict";

    function SendSMSNode(config) {
        RED.nodes.createNode(this, config);
        
        const node = this;
        this.phoneSource = config.phoneSource || "msg";  // msg or fixed
        this.phone = config.phone || "";
        this.messageSource = config.messageSource || "msg";  // msg or fixed
        this.message = config.message || "";
        this.apiKey = config.apiKey || "";
        this.username = config.username || "";

        node.on('input', function(msg) {
            try {
                // Get phone number
                let phone = node.phone;
                if (node.phoneSource === "msg") {
                    phone = msg.phoneNumber || msg.phone || msg.payload.phoneNumber || msg.payload.phone;
                }

                // Get message content
                let messageText = node.message;
                if (node.messageSource === "msg") {
                    messageText = msg.payload || msg.message || msg.sms;
                    if (typeof messageText === 'object') {
                        messageText = messageText.message || messageText.text || "";
                    }
                }

                // Validate inputs
                if (!phone) {
                    node.error("Phone number is required", msg);
                    return;
                }

                if (!messageText) {
                    node.error("Message text is required", msg);
                    return;
                }

                // Log the SMS being sent
                node.log(`Sending SMS to ${phone}: ${messageText}`);
                node.status({fill:"blue", shape:"dot", text:"sending..."});

                // In a real implementation, this would call Africa's Talking API
                // For now, we'll simulate the send
                const result = {
                    phoneNumber: phone,
                    message: messageText,
                    status: "sent",
                    timestamp: new Date().toISOString(),
                    messageId: "msg_" + Date.now()
                };

                // Set output message
                msg.smsResult = result;
                msg.status = "sent";

                node.status({fill:"green", shape:"dot", text:"sent"});
                node.send(msg);

            } catch (err) {
                node.error("Error sending SMS: " + err.message, msg);
                node.status({fill:"red", shape:"ring", text:"error"});
            }
        });

        node.on('close', function() {
            node.status({});
        });
    }

    RED.nodes.registerType("send-sms", SendSMSNode);
};
