// kumva-sms-nodes/nodes/wait-for-response.js
/**
 * WaitForResponse Node for Node-RED
 * Waits for incoming SMS responses via Africa's Talking webhook
 * Routes messages to different outputs based on user input/choices
 */

module.exports = function(RED) {
    "use strict";

    function WaitForResponseNode(config) {
        RED.nodes.createNode(this, config);
        
        const node = this;
        this.timeout = config.timeout || 300; // Default 5 minutes
        this.choices = config.choices || [];
        this.phoneField = config.phoneField || "phoneNumber";
        this.responseField = config.responseField || "payload";
        this.caseInsensitive = config.caseInsensitive !== false;

        // Store active conversations
        const activeConversations = new Map();

        node.on('input', function(msg) {
            try {
                // Get the phone number to wait for response from
                let phoneNumber = msg[node.phoneField] || msg.payload?.phoneNumber;
                
                if (!phoneNumber) {
                    node.error("Phone number is required", msg);
                    return;
                }

                node.log(`Waiting for response from ${phoneNumber}`);
                node.status({fill:"blue", shape:"dot", text:"waiting for response..."});

                // Create a timeout for this conversation
                const conversationId = `${phoneNumber}_${Date.now()}`;
                const timeoutMs = node.timeout * 1000; // Convert to milliseconds

                // Store the message context for this conversation
                activeConversations.set(phoneNumber, {
                    originalMsg: msg,
                    conversationId: conversationId,
                    timeoutHandle: null
                });

                // Set timeout to clean up if no response
                const timeoutHandle = setTimeout(() => {
                    if (activeConversations.has(phoneNumber)) {
                        activeConversations.delete(phoneNumber);
                        const timeoutMsg = Object.assign({}, msg);
                        timeoutMsg.timeout = true;
                        timeoutMsg.status = "timeout";
                        node.status({fill:"red", shape:"ring", text:"timeout waiting for response"});
                        node.send([null, null, timeoutMsg]); // Send to timeout output
                    }
                }, timeoutMs);

                // Update the stored conversation with timeout handle
                const conversation = activeConversations.get(phoneNumber);
                if (conversation) {
                    conversation.timeoutHandle = timeoutHandle;
                }

                // Register this node as a global listener for SMS responses
                if (!global._nodeRedSmsResponses) {
                    global._nodeRedSmsResponses = {};
                }
                global._nodeRedSmsResponses[node.id] = node;

            } catch (err) {
                node.error("Error in wait-for-response: " + err.message, msg);
                node.status({fill:"red", shape:"ring", text:"error"});
            }
        });

        // Handle incoming SMS response
        node.handleSmsResponse = function(phoneNumber, responseText) {
            if (!activeConversations.has(phoneNumber)) {
                return false; // Not waiting for response from this number
            }

            const conversation = activeConversations.get(phoneNumber);
            clearTimeout(conversation.timeoutHandle);
            activeConversations.delete(phoneNumber);

            const originalMsg = conversation.originalMsg;
            const responseMsg = Object.assign({}, originalMsg);
            responseMsg.payload = responseText;
            responseMsg.response = responseText;
            responseMsg.phoneNumber = phoneNumber;

            // Parse response and route to appropriate choice
            const cleanResponse = node.caseInsensitive ? responseText.trim().toUpperCase() : responseText.trim();
            
            let outputIndex = -1;
            for (let i = 0; i < node.choices.length; i++) {
                const choice = node.choices[i];
                const cleanChoice = node.caseInsensitive ? choice.value.toUpperCase() : choice.value;
                
                if (cleanResponse === cleanChoice) {
                    outputIndex = i;
                    responseMsg.choice = choice.label;
                    responseMsg.choiceIndex = i;
                    break;
                }
            }

            if (outputIndex >= 0) {
                node.log(`Response "${responseText}" matched choice index ${outputIndex}`);
                node.status({fill:"green", shape:"dot", text:`received: ${responseText}`});
            } else {
                // Send to last output (unmatched responses)
                outputIndex = node.choices.length;
                node.log(`Response "${responseText}" didn't match any choice`);
                node.status({fill:"yellow", shape:"dot", text:`unmatched: ${responseText}`});
            }

            // Send to appropriate output
            const outputs = new Array(node.choices.length + 2).fill(null);
            outputs[outputIndex] = responseMsg;
            node.send(outputs);

            return true;
        };

        node.on('close', function() {
            // Clean up timeouts
            activeConversations.forEach((conversation) => {
                clearTimeout(conversation.timeoutHandle);
            });
            activeConversations.clear();
            
            // Unregister this node
            if (global._nodeRedSmsResponses && global._nodeRedSmsResponses[node.id]) {
                delete global._nodeRedSmsResponses[node.id];
            }
            
            node.status({});
        });
    }

    RED.nodes.registerType("wait-for-response", WaitForResponseNode);
};
